Update testing chapter (#163)
infinum/frontend-handbook: React/Testing best practices.md

  // or something more meaningful
  it("should call onClick handler", () => {
    const buttonText = "click here";
    const testOnClick = jest.fn();

    render(
      <Button onClick={testOnClick}>{buttonText}</Button>
    );

    user.click(screen.getByText(buttonText));

    expect(testOnClick).toBeCalledTimes(1);
  });
});
```

**Components that uses a base component test example**

Component:

```tsx
import { Button } from "components/Button";

const UserCard: FC<UserCardProps> = ({ title }) => (
  <Card>
    <h3>{title}</h3>
    <Button>click</Button>
  </Card>
);
```

Test:

Here, we mock our `Button` component implementation because we don't care about it here, we only care about the specifics of the `UserCard` component. This is especially useful when you don't want to generate a large tree inside your test for components that don't have any impact on the actual test.

```jsx
import { screen } from '@testing-library/react';

jest.mock("components/button");
(Button as jest.Mock).mockReturnValue(<button />);

describe("UserCard", () => {
  it("should display the correct title", () => {
    const username = "Test User";

    render(<UserCard title={username} />);

>     expect(screen.(username)).toBeDefined();

Reviewer: ```suggestion
    expect(screen.getByText(username)).toBeDefined();
```

As we talked in person, please add a paragraph on using screen over destructuring the render method return object.

      onConfirm = jest.fn();

      render(
        <AlertButton
          title={title}
          confirmButtonText={confirmButtonText}
          onConfirm={onConfirm}
        />
      );
    });

    describe('when clicked', () => {
      beforeEach(() => {
        fireEvent.click(screen.getByText(buttonText));
      });
      
      it('should open alert dialog', () => {
				expect(screen.queryByRole('alertdialog')).toBeNull();
      });
      it('should display correct title', () => {
        expect(screen.queryByText(title)).not.toBeNull();
      });
    });

    describe('when confirm is clicked', () => {
      beforeEach(async () => {
        await waitFor(() => {
          fireEvent.click(screen.getByText(buttonText));
        });
      });
      
      it('should close the dialog', () => {
				expect(screen.queryByRole('alertdialog')).toBeNull();
      });

      // ... more test cases
    });
  })
```

> **Note:** This is a shortened example of this concept, you can refer to the standup video section for more info:
[Next.js testing - Testing shared components](https://youtu.be/KfaFyB0uedk?t=1005)

## Testing user events

> If any part of your test is preforming the state update, that action needs to be wrapped into `waitFor`.

Reviewer: ```suggestion
If any part of your test is performing the state update, that action needs to be wrapped into `waitFor`.
```

You would be also warned for the bad testing practices by the react-testing-library linter plugin.

> For simulating an event, we use `@testing-library/user-event` library.

Reviewer: ```suggestion
For simulating an event, we use the `@testing-library/user-event` library.
```

**Avoid unnecessary "is rendering" test**

Since we have no value in testing whether our component will correctly render, avoid writing these type of tests and focus on more valuable tests instead.

**Avoid destructuring `render` result for querying/finding elements**

It is recommended that you avoid destructuring the `render(...)` result and use `screen` object instead.

> To see more info about [why you should use screen](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen), and other common mistakes in RTL usage, please refer to the [Kents](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library) blog post.

Reviewer: ```suggestion
To see more info about [why you should use screen](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen), and other common mistakes in RTL usage, please refer to [Kent's](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library) blog post.
```

export default UserPage;
```

> In this page example we should test all three states of the page component. To do that we should mock `<ErrorPage>`, `<LoadingPage>` and `<UserTemplate>` and check if it's rendered based on fetcher response.

Reviewer: ```suggestion
In this page example we should test all three states of the page component. To do that we should mock `<ErrorPage>`, `<LoadingPage>` and `<UserTemplate>` and check if it's rendered based on the fetcher response.
```

### SWR testing

> To use swr in your tests, add `SWRConfig` to your `AllProviders` mock with following setup (note: your cases could require more customization):

Reviewer: ```suggestion
To use SWR in your tests, add `SWRConfig` to your `AllProviders` mock with the following setup (note: your cases could require more customization):
```

```tsx
const AllProviders = ({ children }) => (
  <SWRConfig value={{ dedupingInterval: 0, provider: () => new Map() }}>
    <ChakraProvider theme={theme}>{children}</ChakraProvider>
  </SWRConfig>
);
```

> With this setup, we are sure that each component we're testing will have it's own cache, so we don't have to manually clear cache before each test.

Reviewer: ```suggestion
With this setup, we are sure that each component we're testing will have its own cache, so we don't have to manually clear the cache before each test.
```
